### ISSUE TITLE
It only works when the view validation page is open.

### ISSUE THREAD

**Issue by midorigame737** [csapuntz/siteblock#8]
[After translation]
I'm not good at English, so I use Google Translate.
The site will only be blocked if the JS code is opened in the extension's Verify page view.
Also, restarting Chrome will reset the browsing time and the site will not be blocked again. (Is this a specification?)
environment
OS:Win11
Browser: Google Chrome
[Original text]
英語が苦手なため、Google翻訳を使用しています。
拡張機能のページのビューを検証でJSのコードを開いている状態でしかサイトのブロックが行われません。
また、Chromeを再起動すると閲覧時間がリセットされ再びサイトがブロックされない状態になります。（こちらについては仕様でしょうか?）
環境
OS:Win11
ブラウザ:GoogleChrome

**Comment by csapuntz** [comment]
Thanks for bug report! I will attempt to reproduce the problem and come up with a fix.

**Comment by csapuntz** [comment]
Hi, I am unable to reproduce the problem. 

Would you be willing to try adding a new empty, local chrome profile and using Siteblock as the only extension in that profile? Do you have the same problem? Can you send a screenshots or a video?

**Comment by keroberotte** [comment]
[After translation]

I have been able to reproduce the problem, so I will share the steps to reproduce it and the video.

1. Make time_used visible from the options screen (I have attached html/options.html and js/display.js for this purpose)
2. Install siteblock
3. Add to restricted sites and browse for more than one minute (keep using the mouse to avoid Chrome going into idle mode)
4. Go to the options screen and stop using the PC for more than one minute
5. Press the refresh button and time_used will return to 0

I suspect that the main cause of the problem is that background.js is re-executed when Chrome returns from idle mode, but I am not sure of the specific process by which time_used is reset.
Thank you in advance.

[Original text]
問題の再現ができたため、再現手順と動画を共有します。

1. time_used を options画面から確認可能にする（このための html/options.html, js/display.js を添付します）
2. siteblock をインストールする
3. 制限サイトに追加し、1分以上閲覧する（このとき、Chromeがアイドル状態になるのを避けるためマウス操作を続ける）
4. options画面に移動し、1分以上PCの操作をやめる
5. 更新ボタンを押すと、time_usedが0に戻っている

Chromeがアイドル状態から復帰した際に background.js が再実行されるのが問題の主原因だと推測していますが、time_usedがリセットされる具体的な流れはよくわかっていません。
よろしくお願いいたします。

https://github.com/user-attachments/assets/4211f821-d234-4fbd-bb42-f4e2ea0bcb3c

[assets.zip](https://github.com/user-attachments/files/19278582/assets.zip)

**Comment by csapuntz** [comment]
Yeah, this page explains it - https://developer.chrome.com/docs/extensions/develop/migrate/to-service-workers - the service workers get stopped and siteblock shouldn't be using global variables. 

I was in too much of a rush when I switched to manifest v3.

**Comment by csapuntz** [comment]
Ok, I wasn't using global variables, but I think I forgot an async/await. Much gratitude if you are able to test 6b0e09d.

**Comment by keroberotte** [comment]
[After translation]
Thank you for the quick fix. I appreciate it because I don't think I could fix it myself.

However, I tested 6b0e09d and it doesn't seem to fix the problem.

I'm a beginner with Chrome extensions, so I'm sorry if I'm misunderstanding something, but isn't `var sb = csapuntz.siteblock.newSiteBlock();` in js/background.js a global variable?

**Comment by csapuntz** [comment]
Ok, more fixes in f49e2bf0ca.

**Comment by csapuntz** [comment]
@keroberotte thank you so much for the testing on your end!

**Comment by keroberotte** [comment]
[After translation]
Thank you for the fix.
However, it looks like the problem persists :cry:
I appreciate you taking the time to fix it.

**Comment by keroberotte** [comment]
[After translation]
Thank you for the fix.
It seems my checking method was incorrect, and even though time_used had returned to 0, the site continued to be blocked.
I appreciate you taking your valuable time to fix it.

I'll try using it for a while!

**Comment by keroberotte** [comment]
[After translation]
I have deleted the message several times. I'm sorry for the confusion...

Thank you for the fix.

However, the problem seems to remain.

After the storage time_used reaches 0, the block is released if the PC is left idle for one minute several times. (Video attached)

https://github.com/user-attachments/assets/9d703b55-b322-40af-9136-bc5e9c7ca3e7

Thank you for taking the time to fix the problem.

**Comment by csapuntz** [comment]
The service worker shutdown tears down state needed to do usage tracking (like endfunc, tabStart, time_start). Will need to figure out another way to track usage that works.

**Comment by csapuntz** [comment]
Latest commit (5d4d992b68f9caebd75b9) in the repository persists the list of tabs, start time, end time, and tab used. I think it should behave more predictably across service worker restarts. 

One problem with the current code is that it will persist across Chrome shutdown and restarts too - so the usage timer will be running even Chrome is off or the extension is disabled!

Will work on a commit to clean up the trackedTabs on startup. Also probably need to not accumulate timeUsed if it took > 2 minutes or so between alarms

**Comment by keroberotte** [comment]
[After translation]
Thank you for fixing it!
So if I close Chrome while a restricted site is in a tab, it can be blocked unintentionally. I understand the problem.

I don't think there's any problem with the way I use it, so I'll keep using 5d4d992 for a while!

**Comment by csapuntz** [comment]
c80216b1 should deal with extension restart and suspend better.

**Comment by keroberotte** [comment]
[After translation]
Thank you for the fix!

Because of the 2-minute alarm gap ignore, time_used couldn't be accumulated accurately for example, when watching YouTube without touching the PC. This was inconvenient for me, so I removed the 2-minute alarm gap ignore and used the extension that way.
However, removing the 2-minute alarm gap ignore caused the time_used reset issue to reoccur after several minutes of inactivity.

Therefore, I thought the problem was that Chrome events were firing before init() completed, so I changed it to ensure init() runs first.
The above fix is ​​open as a PR, please feel free to merge it if you find it helpful.

This gives a very similar experience to the old site block from Manifest V2.
I used it for a day and haven't had any particular issues.
It seems to me that the problem described in the issue has been resolved.

I appreciate your patient efforts in making the fix.
I'm rooting for your continued success.
